client.go

```diff
@@ -615,11 +619,55 @@ func (c *Client) doWithEndpoint(ctx context.Context, endpoint, method, path stri
 		}
 	}
 
-	resp, err = c.HTTPClient.Do(req)
+	resp, err = c.doWithRateLimit(req)
 
 	return c.checkResponse(resp, err)
 }
 
+// handleRatelimitError will handle rate limit errors from responses with http
+// code 429. Delaying retry based on ratelimit-reset recommended by PagerDuty
+// https://developer.pagerduty.com/docs/72d3b724589e3-rest-api-rate-limits#reaching-the-limit
+func (c *Client) doWithRateLimit(req *http.Request) (*http.Response, error) {
+	resp, err := c.HTTPClient.Do(req)
+
+	if err != nil || resp.StatusCode != http.StatusTooManyRequests {
+		return resp, err
+	}
+
+	// Rate limited, wait and retry
+	var delay time.Duration
+	rateLimitReset := resp.Header.Get("ratelimit-reset")
+	resetTime, err := strconv.ParseInt(rateLimitReset, 10, 64)
```

Comment:
I think you should rename `err` in this line to something else. Currently in the case that `rateLimitReset` is an invalid integer, the user will lose the `url.Error` returned by `c.HTTPClient.Do(req)` which provides more useful information and is more common to be handled explicitly, which also adds the possibility of breaking a client if he assumed Do always returns an `url.Error`.
```
resp, err := client.Something()
if err != nil {
    if urlErr := err.(url.Error); urlErr.Timeout() { // This will panic
        // handles timeout...
    }
}

Reply:
Oh Thank you! You are right. Good catch @cjgajard